Update Dockerfile build instructions, add docker-compose file, add dev env docs

nlMeminger opened the pull request.

Commit:
adding script to start the app in a container for development

nlMeminger committed:
temp fix

add a temp fix to mysqlclient not properly installing, and the django server not starting properly.

nlMeminger committed:
changing port 80 to 80000

nlMeminger committed:
adding ingrid as allowed host in development

justwheel commented on Dockerfile:
```diff
@@ -6,8 +6,12 @@ COPY . /usr/src/app
+    && pip3 install pipenv
+
+RUN pip3 install mysqlclient
+RUN pipenv install --system --deploy --dev
```
Since @leong96's changes in #17 were merged, `mysqlclient` is now in the environment. This will work as it was originally written now.

Normally it is best practice to [combine multiple `RUN` steps together](https://rock-it.pl/how-to-write-excellent-dockerfiles/), but here I agree with your change to split APT upgrades and `pipenv` step in order to speed up container build time.

nlMeminger replied:
the pip3 install here is a temporary fix in order to move on from pipenv throwing errors when trying to install mysqlclient.  This will be reversed when time permits before the end of the semester.

justwheel replied:
@nlMeminger Argh. This is annoying. Same thing on my machine too with the dependency. It must not be getting installed correctly. You're right, we should revisit later but I agree to prioritize getting a dev environment done first.

justwheel commented on Dockerfile:
```diff
@@ -6,8 +6,12 @@ COPY . /usr/src/app
 EXPOSE 8000
-CMD ["python3", "manage.py", "runserver", "0.0.0.0:8000"]
+
+RUN sleep 10
```
I don't understand this change. Why run `sleep 10` inside the container?

nlMeminger replied:
Was a testing statement that doesn't hurt anything, but should be removed.

justwheel replied:
@nlMeminger No worries. Let's remove it before this PR merges though, it will be annoying to wait an extra ten seconds for the container to build.

justwheel commented on Dockerfile:
```diff
@@ -6,8 +6,12 @@ COPY . /usr/src/app
-CMD ["python3", "manage.py", "runserver", "0.0.0.0:8000"]
+
+RUN sleep 10
+CMD ["python3", "/usr/src/app/manage.py", "runserver", "0.0.0.0:8000"]
```
The path to `manage.py` should not be hard-coded; it should exist relative to the current working directory, set earlier in the Dockerfile with the `WORKDIR` option. In this case, the change is not necessary.

nlMeminger replied:
The full path should not be necessary, however I was testing an issue where the django app was not starting on on docker-compose up.  The full path can/should be removed but is not necessary at this moment

justwheel commented on docker-compose.yml:
```diff
@@ -7,11 +7,10 @@ services:
     ports:
-      - 80:80
+      - 8000:8000
       - 443:443
```
This mixes a standard port with a custom port. I suggest dropping `443:443` since we are only serving the application via port `8000`. We can drop `443` since it is not needed anymore.

nlMeminger replied:
443 should be opened at a later time so that we can handle HTTPS traffic, and 8000 is being used because our test servers, ie my personal servers, have other services running on port 80.  8000 is the Django default port.

justwheel replied:
In our case, Django serves the application out of one port. We handle HTTP/HTTPS at the web server level in a real deployment environment and can switch between that. From a development POV, Django only binds to a single assigned port.

For example, we could do…

`- 8000:80`

…since this is only used for local development at the moment. HTTPS in development is not required.

justwheel committed:
docs: Reformat / expand on dev environment doc

This commit  builds on @nlMeminger's initial steps and adds more Markdown syntax to make it easier to read. I also added in some other steps to make it easier for other team members to understand these instructions depending on their operating system.

justwheel commented on up.sh:
```diff
@@ -0,0 +1,7 @@
+sudo docker-compose build && sudo docker-compose up -d
+#Display the list of runnng containers for verification
+sudo docker ps
+
```
Could you please add either a toggle flag to this script for not using detached mode, or create an alternate script that builds the container and runs it in `stdout`? Seeing the logs in my terminal is more useful for me while debugging.

nlMeminger replied:
This script is still in development.

justwheel replied:
@nlMeminger Okay. This also requires running Docker as root though. This script might work in your environment, but it will not in others'. Something more general-purpose would be useful. We can talk about this one during the team meeting today.

justwheel reviewed (changes requested):
Hey @nlMeminger, changes look good. I just took a look through the PR. A few more changes are needed before this is ready to merge.

nlMeminger committed:
Update Dockerfile

nlMeminger committed:
Delete up.sh

nlMeminger committed:
Remove port 443 and change port 8000 to 80

justwheel committed:
Use WORKDIR path to start Django application

This commit changes the path used to start the Django application to use the default path specified in `WORKDIR`. When the `WORKDIR` is set, this is the relative path used for all commands in a Dockerfile. So if for some reason, later on you need to change the path, you only need to change the `WORKDIR` instead of in several files.

justwheel reviewed (approved):
Hey @nlMeminger, changes look good. Merging. :ocean:
